Feat/vaults

userdarius opened the pull request.

userdarius committed:
init vault draft

userdarius committed:
remove eject method and update dust requirements

userdarius committed:
new vaults design

userdarius committed:
added some tests

userdarius committed:
improved certain methods + added some tests

ronanyeah committed:
Format w/ indent

ronanyeah committed:
Format w/o indent

ronanyeah committed:
Fix build error + syntax changes

ronanyeah committed:
Remove suilend wrapper functions

ronanyeah committed:
Add liquidity check to withdraw

ronanyeah committed:
Add FeeType enum

ronanyeah committed:
Wire up utilization rate updating

ronanyeah committed:
Use vector macros

ronanyeah committed:
Apply management fee on withdrawal

ronanyeah committed:
Add note

ronanyeah committed:
Use u128 for NAV calculations

ronanyeah committed:
Remove user objects, apply mgmt+perf fees to NAV

ronanyeah committed:
Add oracle share pricing

ronanyeah committed:
Use prod deposit/withdraw functions for tests

ronanyeah committed:
Add allocate/divest manager functions

ronanyeah committed:
Add allocation test

ronanyeah committed:
Add support for multiple lending markets + obligations

ronanyeah committed:
Fix tests

ronanyeah committed:
Improve logic

ronanyeah committed:
Add TS SDK

0xLendlord commented on contracts/vaults/sources/vault.move:
```diff
@@ -0,0 +1,947 @@
+    performance_fee_bps: u64,
+    deposit_fee_bps: u64,
+    withdrawal_fee_bps: u64,
+    treasury_cap: TreasuryCap<VaultShare<P>>,
```
I think this will not work becauase `TreasuryCap<VaultShare<P>>` requires the VaultShare type to be OTW if I am not mistaken. Perhaps an easy way to test this is to have a coin type module in the tests and have it create treasury cap without the use of test_only functions

0xLendlord replied:
if this is the case we have to options:

- keep `VaultShare` wrapper type and create Supply inner object instead of TreasuryCap (pro: better devX, con: cannot create token metadata)
- remove `VaultShare` (con: worst devX, pro: token has metadata)

0xLendlord commented on contracts/vaults/sources/vault.move:
```diff
@@ -0,0 +1,947 @@
+    mut deposit: Coin<T>,
+    lending_market: &LendingMarket<L>,
+    clock: &Clock,
+    agg: VaultValueAggregate,
```
what is the reasoning for creating the vaultAggregate as a separate function instead of inside the deposit / withdraw function?

ronanyeah committed:
Fix some calculations + tests

ronanyeah committed:
type_name::get -> with_defining_ids

ronanyeah committed:
Use balance::Supply for vault shares

ronanyeah committed:
Change vault share types

ronanyeah committed:
Update deps

ronanyeah committed:
Regenerate TS sdk

ronanyeah committed:
Refactor

ronanyeah committed:
Use Obligation deposited_value_usd + unweighted_borrowed_value_usd

ronanyeah committed:
Tidy up tests,  add NAV test

ronanyeah committed:
Doc

ronanyeah committed:
Consolidate fees

ronanyeah committed:
Add product doc

ronanyeah committed:
Tidy up utilization rate usage

ronanyeah committed:
Add notes

ronanyeah committed:
Adjust perf fees

ronanyeah committed:
Adjust user withdraw

ronanyeah committed:
Publish

ronanyeah committed:
Add compound test

ronanyeah committed:
Use unique currencies for vault shares

ronanyeah committed:
Publish

ronanyeah committed:
Add compounding functions

ronanyeah committed:
Restrict compound w/ swap to vault manager

ronanyeah committed:
Add version module

ronanyeah committed:
Remove unused

ronanyeah committed:
Reorganise functions

ronanyeah committed:
Add VaultStats event

ronanyeah committed:
Add error attributes + messages

ronanyeah committed:
Change pyth handling + publish

ronanyeah committed:
Add total shares to stats event

ronanyeah committed:
Get base token decimals from reserve

ronanyeah committed:
Add error codes to fail tests

ronanyeah committed:
Error changes

ronanyeah committed:
Add helper

ronanyeah committed:
Add back LendingMarket as price source

ronanyeah committed:
Publish

ronanyeah committed:
Add segmented allocation data to stats event

ronanyeah committed:
Fix share calculation

ronanyeah committed:
Move assert

ronanyeah committed:
Enforce minimum deposit as USD

ronanyeah committed:
Add more info to deploy/divest events

ronanyeah committed:
Convert various u64 usage to Decimal

ronanyeah committed:
Publish

ronanyeah committed:
Adjust test

ronanyeah committed:
Remove utilization cap

ronanyeah committed:
More validation

ronanyeah committed:
Change agg params to references

ronanyeah committed:
Add reward crank structure

ronanyeah committed:
Fee adjustments

ronanyeah committed:
Publish

ronanyeah committed:
Use market value price

ronanyeah committed:
Assert price freshness

ronanyeah committed:
Add swap token type guard

ronanyeah committed:
Add Vault.slippage_bps and use in swap

ronanyeah committed:
LM param docs

ronanyeah committed:
Name change

ronanyeah committed:
Fix doc

ronanyeah committed:
Add fee helper

ronanyeah committed:
Refactor deposit

ronanyeah committed:
Add price freshness checks

ronanyeah committed:
Use USD market price

ronanyeah committed:
Update vault crank + add test

ronanyeah committed:
Add lending unwind

ronanyeah committed:
Make reward swap compound permissionless

ronanyeah committed:
Add import

ronanyeah committed:
Add price test, change function visibility

ronanyeah committed:
Refactor value aggregate updates

ronanyeah committed:
Move functions

ronanyeah committed:
Add vault metadata

ronanyeah committed:
Adjust type params

ronanyeah committed:
Ensure complete unwind

ronanyeah committed:
Refactor manager fee accrual, add tests

ronanyeah committed:
Remove comments

ronanyeah committed:
Add vault share validation

ronanyeah committed:
Update product doc

ronanyeah committed:
Remove unused public functions

ronanyeah committed:
Allow unwind no-op

ronanyeah committed:
Add event emit

ronanyeah committed:
Publish

ronanyeah committed:
Various refactoring + features

- create accumulator + utils modules
- rename events
- add unique ticket for aggregation process
- allow managers to withdraw and swap rewards

ronanyeah committed:
Reduce fee caps

ronanyeah committed:
Add minimum crank interval

ronanyeah committed:
Update share metadata

ronanyeah committed:
Adjust errors

ronanyeah committed:
Add icon url check

ronanyeah committed:
Comments  + function rename

ronanyeah committed:
Improve test

ronanyeah committed:
Don't enforce reward claim for non-oracle tokens

ronanyeah committed:
Remove steamm reward swap

ronanyeah committed:
Change mgmt fee cap

ronanyeah committed:
Update docs

ronanyeah committed:
Use contract const for slippage

ronanyeah committed:
Add vaults to Makefile

ronanyeah committed:
Publish

ronanyeah committed:
Publish

ronanyeah committed:
Doc comments

ronanyeah committed:
Slippage doc change

ronanyeah committed:
Update tests to use TestRunner

ronanyeah committed:
Add vault version migrate to crank

ronanyeah committed:
Edit param name

ronanyeah committed:
Compound rewards to vault

ronanyeah committed:
Remove view function

ronanyeah committed:
Test updates

ronanyeah committed:
Add U64_MAX to fee claim

ronanyeah committed:
Edit comment

ronanyeah committed:
Handle SUI unstaking

ronanyeah committed:
Update deps

ronanyeah committed:
Publish

ronanyeah committed:
Product doc update

ronanyeah committed:
Allow manager to re-order obligations for unwind

ronanyeah committed:
Refactor withdraw fee accrual

ronanyeah committed:
Prevent reward compounding during accumulation

ronanyeah committed:
Refactor method calls

ronanyeah committed:
Don't factor mgmt fee shares into perf fees

ronanyeah committed:
Fix high-water-mark initialization for perf fees

ronanyeah committed:
Only enforce reward compound if >1USD and in MAIN_POOL

ronanyeah committed:
Add obligation refresh to crank

ronanyeah committed:
Only refresh active obligations

ronanyeah committed:
Redesign unwind process

ronanyeah committed:
Add min swap zero check

ronanyeah committed:
Abort on unexpected borrows

ronanyeah committed:
Only allow passing active LendingMarkets as oracle sources

ronanyeah committed:
Track trusted lending market as price source

ronanyeah committed:
Clarify test currency names

ronanyeah committed:
Publish
